File: tools-manage/invocationlab-admin/src/main/java/org/javamaster/invocationlab/admin/aspect/ErdRolesAllowedAspect.java
package org.javamaster.invocationlab.admin.aspect;

import org.javamaster.invocationlab.admin.annos.ErdRolesAllowed;
import org.javamaster.invocationlab.admin.config.ErdException;
import org.javamaster.invocationlab.admin.enums.RoleGroupEnum;
import org.javamaster.invocationlab.admin.model.erd.TokenVo;
import org.javamaster.invocationlab.admin.service.ErdOnlineRoleService;
import com.google.common.collect.Sets;
import org.apache.commons.lang3.tuple.Pair;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpSession;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.Collections;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * @author yudong
 */
@Aspect
@Component
public class ErdRolesAllowedAspect {
    @Autowired
    private ErdOnlineRoleService erdOnlineRoleService;

    @SuppressWarnings({"DataFlowIssue", "ConstantConditions"})
    @Around("@annotation(org.javamaster.invocationlab.admin.annos.ErdRolesAllowed)")
    public Object checkUserRole(ProceedingJoinPoint joinPoint) throws Throwable {
        ServletRequestAttributes requestAttributes = (ServletRequestAttributes) RequestContextHolder.getRequestAttributes();
        @SuppressWarnings("DataFlowIssue")
        HttpServletRequest request = requestAttributes.getRequest();
        String projectId = request.getHeader("projectId");
        HttpSession session = request.getSession();
        TokenVo tokenVo = (TokenVo) session.getAttribute("tokenVo");

        Pair<Set<Long>, RoleGroupEnum> pair = erdOnlineRoleService.getUserRoleIds(tokenVo.getUserId(), projectId);
        Set<Long> userRoleIds = pair.getKey();
        Object[] args = joinPoint.getArgs();
        MethodSignature signature = (MethodSignature) joinPoint.getSignature();
        Method method = signature.getMethod();
        ErdRolesAllowed rolesAllowed = method.getAnnotation(ErdRolesAllowed.class);
        Set<Long> methodRoleIds = Arrays.stream(rolesAllowed.value())
                .map(roleEnum -> roleEnum.id)
                .collect(Collectors.toSet());
        if (Collections.disjoint(Sets.newHashSet(userRoleIds), methodRoleIds)) {
            throw new ErdException(rolesAllowed.msg());
        }
        return joinPoint.proceed(args);
    }

}

File: tools-manage/invocationlab-admin-21/src/main/kotlin/org/javamaster/invocationlab/admin/aspect/ErdRolesAllowedAspect.kt
package org.javamaster.invocationlab.admin.aspect

import org.javamaster.invocationlab.admin.annos.ErdRolesAllowed
import org.javamaster.invocationlab.admin.config.ErdException
import org.javamaster.invocationlab.admin.enums.RoleEnum
import org.javamaster.invocationlab.admin.model.erd.TokenVo
import org.javamaster.invocationlab.admin.service.ErdOnlineRoleService
import org.javamaster.invocationlab.admin.util.SpringUtils
import com.google.common.collect.Sets
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.reflect.MethodSignature
import org.springframework.stereotype.Component
import org.springframework.web.context.request.RequestContextHolder
import org.springframework.web.context.request.ServletRequestAttributes
import java.util.*
import java.util.stream.Collectors

/**
 * @author yudong
 */
@Aspect
@Component
class ErdRolesAllowedAspect {

    @Around("@annotation(org.javamaster.invocationlab.admin.annos.ErdRolesAllowed)")
    fun checkUserRole(joinPoint: ProceedingJoinPoint): Any {
        val requestAttributes = RequestContextHolder.getRequestAttributes() as ServletRequestAttributes
        val request = requestAttributes.request
        val projectId = request.getHeader("projectId")
        val session = request.session
        val tokenVo = session.getAttribute("tokenVo") as TokenVo

        val erdOnlineRoleService = SpringUtils.context.getBean(ErdOnlineRoleService::class.java)
        val pair = erdOnlineRoleService.getUserRoleIds(tokenVo.userId!!, projectId)
        val userRoleIds = pair.key
        val args = joinPoint.args
        val signature = joinPoint.signature as MethodSignature
        val method = signature.method
        val rolesAllowed = method.getAnnotation(ErdRolesAllowed::class.java)
        val methodRoleIds = Arrays.stream(rolesAllowed.value)
            .map { roleEnum: RoleEnum -> roleEnum.id }
            .collect(Collectors.toSet())
        if (Collections.disjoint(Sets.newHashSet(userRoleIds), methodRoleIds)) {
            throw ErdException(rolesAllowed.msg)
        }
        return joinPoint.proceed(args)
    }
}
